### DISCUSSION TITLE
[version constraints] Make Accept-Version header optional

### DISCUSSION THREAD

**Discussion by aheckmann** [fastify/fastify#5649] [Q&A] [answered]
I'm considering using the built-in [version constraints](https://fastify.dev/docs/latest/Reference/Routes/#version-constraints) and wondering why the `accept-version` header is required when using them. I'd like to enable versioning for an existing API but if I use the built-in versioning implementation, I will break existing API clients who do not already set the `accept-version` header.

- What is the reason for fastify replying with a 404 when the `accept-version` header is not set?

Barring a solid reason to the above question, I may contribute a PR to modify this behavior so existing clients are not broken when adding versioning to an existing API.

**Comment by Eomm** [comment]
> What is the reason for fastify replying with a 404 when the accept-version header is not set?

It is due a missing route without constraints.

> Barring a solid reason to the above question, I may contribute a PR to modify this behavior so existing clients are not broken when adding versioning to an existing API.

We may create a new option to customize the header name, or you may use this utility https://github.com/Eomm/header-constraint-strategy to create your `client-version` constraint too.

**Reply by aheckmann** [reply]
The route is not missing though. From a client perspective, the fastify API will be suddenly broken. I'd like fastify versioning, out of the box, to support enabling versioning without it being a breaking change. I know I can roll my own constraints or use a community plugin but this post is about improving Fastify itself so that we can enable versioning without breaking clients.

**Reply by Eomm** [reply, marked as answer]
Could you provide a snippet?

This code is working as expected:

```js
const app = require('fastify')({ logger: true });
app.get('/', { constraints: { version: '1.0.0' } }, async () => 'v1');
app.get('/', { constraints: { version: '2.0.0' } }, async () => 'v2');
app.get('/', async () => 'v2'); // default version

app.inject('/').then((res) => {
  console.log('v2: ', res.body);
});
app
  .inject({
    url: '/',
    headers: { 'accept-version': '1.0.0' },
  })
  .then((res) => {
    console.log('v1: ', res.body);
  });
app
  .inject({
    url: '/',
    headers: { 'accept-version': '2.0.0' },
  })
  .then((res) => {
    console.log('v2 header: ', res.body);
  });

```

**Reply by Eomm** [reply]
> to support enabling versioning without it being a breaking change

It is a dev choice to say: "if the header is missing, use the older (or the newer) logic" TBH

Regarding the `accept-version`, I think it is ok for the user to customize it with a new option

**Reply by aheckmann** [reply]
Ah that makes sense. I didn't realize I needed to register the route again to enable the fallback behavior. Thank you.
